src/clients/Deluge.ts

```diff
@@ -0,0 +1,236 @@
+	}
+
+	/**
+	 * validates the login and host for deluge webui
+	 */
+	async validateConfig(): Promise<void> {
+		const url = new URL(this.delugeWebUrl);
+		if (url.username && !url.password) {
+			throw new CrossSeedError(
+				"you need to define a password in the delugeWebUrl. (eg: http://:<PASSWORD>@localhost:8112)"
+			);
+		}
+		const result = await this.authenticate();
+		if (!result) {
+			logger.verbose({
+				label: Label.DELUGE,
+				message: "failed to authenticate the deluge daemon.",
+			});
+			throw new CrossSeedError(
+				`failed to authenticate the deluge daemon.`
+			);
+		} else {
+			this.enLabeled = this.labelEnabled();
+		}
+	}
+
+	/**
+	 * connects and authenticates to the webui
+	 */
+	private async authenticate(): Promise<boolean> {
+		if (!this.loggedIn) {
+			let response = await this.call({
+				params: [this.delugeWebUrl.password],
+				method: "auth.login",
+			});
+			if (response === null) {
+				throw new CrossSeedError(
+					`Could not reach deluge: ${this.delugeWebUrl.origin}`
+				);
+			}
+			if (
+				response &&
+				response.headers &&
+				response["data"]["result"] &&
+				response.headers["set-cookie"]
+			) {
+				this.delugeCookie =
+					response.headers["set-cookie"][0].split(";")[0];
+				this.loggedIn = true;
+			}
+		} else {
+			let response = await this.call({
+				params: [this.delugeCookie],
+				method: "auth.check_session",
+			});
+			this.loggedIn = !!response;
+		}
+		return this.loggedIn;
+	}
+
+	/**
+	 * authenticates and sends JSON-RPC calls to deluge
+	 */
+	private async sendCall(method: string, params: any[]) {
+		let response;
+		if (!this.loggedIn) {
+			await this.authenticate();
+			response = await this.call({
+				params,
+				method,
+			});
+		} else
+			response = await this.call({
+				params: params,
+				method,
+			});
+
+		return response && response.data && response.data.error === null
+			? response.data
+			: response.data.error;
+	}
+
+	private async call(body: any): Promise<AxiosResponse | null> {
+		return new Promise((resolve) => {
+			body.id = ++this.msgId;
+			if (this.msgId > 1024) this.msgId = 0;
+			axios
+				.post(
+					this.delugeWebUrl.origin + this.delugeWebUrl.pathname,
+					body,
+					{
+						headers: {
+							"Content-Type": "application/json",
+							Cookie: this.delugeCookie,
+						},
+					}
+				)
+				.then((data) => {
+					resolve(data);
+				})
+				.catch(() => resolve(null));
+		});
+	}
+
+	/**
+	 * lists plugins and adds/sets labels
+	 * returns true if successful.
+	 */
+	private async labelEnabled() {
+		const enabledLabels = await this.sendCall(
+			"core.get_enabled_plugins",
+			[]
+		);
+		return enabledLabels["result"].includes("Label");
+	}
+	private async setLabel(torrenthash: string) {
+		if (this.enLabeled) {
+			let setResult = await this.sendCall("label.set_torrent", [
+				torrenthash,
+				this.delugeLabel,
+			]);
+			if (
+				setResult["message"] &&
+				setResult["message"].includes("Unknown Label")
```

Comment:
is there like an `isError` boolean we can check?

Reply:
setResult is actually like a full stacktrace (in message) and a code that does I can't seem to locate any documentation on.

Normally, this would be done via a UI (either native or webui) where setting erroneous labels is impossible, so this was likely not taken into account. It does not crash anything though.

setResult response below.

```
{
  message: "Failure: [Failure instance: Traceback (failure with no frames): <class 'deluge.error.WrappedException'>: Unknown Label\n" +
    'Traceback (most recent call last):\n' +
    '  File "/usr/lib/python3.11/site-packages/deluge/core/rpcserver.py", line 334, in dispatch\n' +
    '    ret = self.factory.methods[method](*args, **kwargs)\n' +
    '          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n' +
    '  File "/usr/lib/python3.11/site-packages/deluge/plugins/Label-0.3-py3.11.egg/deluge_label/core.py", line 317, in set_torrent\n' +
    "    check_input((not label_id) or (label_id in self.labels), _('Unknown Label'))\n" +
    '  File "/usr/lib/python3.11/site-packages/deluge/plugins/Label-0.3-py3.11.egg/deluge_label/core.py", line 65, in check_input\n' +
    '    raise Exception(message)\n' +
    'Exception: Unknown Label\n' +
    '\n' +
    ']',
  code: 4
}
```